### DISCUSSION TITLE
Update branch with rebase repository setting

### DISCUSSION THREAD

**Discussion by avaly** [community/community#12032] [Pull Requests]
The new [update branch](https://github.blog/changelog/2022-02-03-more-ways-to-keep-your-pull-request-branch-up-to-date/) feature is great!

But for teams which are only using the rebase functionality, it's not optimal to always have to pick the rebase option from a drop down menu.

It would be ideal if there were a repository level setting to configure which option is the default for that button.

**Comment by michaelbeaumont** [comment]
It'd be great for it to even just remember the last used option like the "Merge PR" option does.

**Reply by filipesperandio** [reply]
💯  even better then having a default.

**Comment by burz** [comment]
The issue I am facing is that even when merge commits are disabled in the repository settings, the update button still defaults to merge commits.

Please change the "update branch" button to either follow the same rules as the overall PR options, or allow it to be configured separately to disallow certain options.

(The "update branch" is a great feature, btw!)

**Reply by rmacklin** [reply]
I would vote for separate configuration as this would allow for the button to be used even when the project desires ["semi-linear" history](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#merge-commit-with-semi-linear-history) (where merge commits are allowed to be created when merging a fully-rebased branch into the mainline).

**Comment by rmacklin** [comment]
This was also requested in a comment on another feature request: https://github.com/community/community/discussions/3245#discussioncomment-2105732 and that comment currently has more upvotes than this post (so the upvotes here are under-represented)

**Comment by ManuelRauber** [comment]
That would be an awesome feature, if we choose the strategy to update the branch, just like we do with PR merging.

**Comment by rathpc** [comment]
Hey @scotttesler, any chance we can get some traction on this please? 🙏

**Comment by dimisjim** [comment]
doubtful, with all the layoffs circling around.

**Comment by marcusramberg** [comment]
@krisnova ^^ This is the issue I was talking about on your Twitch stream. 😸

**Comment by ferologics** [comment]
ditto

**Comment by valentin-krasontovitsch** [comment]
there's no way for a user (non github employee) to contribute to the github codebase and make this happen, is there?

**Comment by deefactorial** [comment]
Is there any update from the GitHub team regarding this feature request?

**Comment by dcornewell** [comment]
We were just discussing how update with rebase is preferred but most of my dev team just clicks the button. I'd like it at least to be configurable to be the first option to save a click.

**Comment by GANTR0N** [comment]
This would be a major help for how our testing system is integrated, this is a small change that would be a big QoL help.

**Comment by williamglasse** [comment]
Our team would really benefit from this too! Have had a couple of team members accidentally merge, and then have had to undo that change. Would really like to be able to select a default other than "merge"

**Comment by mbirchdk** [comment]
Would love this feature! Sucks when teammembers accidentally presses the merge button

Personally I want merge-commits from feature-branches on to the main-branch, but I dont want to allow merge-commits for updating other branches, this should be done using rebase. I tried enforcing it with the branch-protection rule "require linear history" on all other branches than main. But unfortunately it uses the entire commit-history, and not just the commits on the specific branch

**Comment by manemarron** [comment]
Our team would love this feature as well. We would like the feature branches to be updated using rebase to avoid having so many merge-commits. Our master, staging, and develop branches would be updated using merges obviousy, but for feature branches or fix branches, this feature would help us have a cleaner commit history in a simpler way.

**Comment by minherz** [comment]
I guess the only way is to contribute this feature because Github team is clearly ignorant about it.

**Reply by jajik** [reply]
How? GitHub is not an open source project.

**Reply by DomasM** [reply]
Yeah there is no way to do that. Possible workaround would be to use frontend hacking e.g. Tampermonkey to edit Github UI, but of course that's quite nasty

**Comment by sandstrom** [comment]
Yes, this is annoying, causes a lot of frustration when people click this by accident.

Our workflow is to require 'merge commits' when merging a PR, but we only use 'rebase' when updating a branch. In other words, no merges *from* master into a feature branch.

**Comment by dedalus-adb** [comment]
This feature is so much needed, just like the comment in https://github.com/orgs/community/discussions/3245#discussioncomment-2105732 said. Our team works exactly lilke this, always asking for rebases instead of merges for updating branches, and it would be so easy to just be able of configure the default behaviour of the button, just like we can configure the behaviour of the "Merge/Squash and Merge/Rebase" button...

I hope this can be implemented in the near future!

**Comment by robertIsaac** [comment]
it would be much helpful if the admins has the choice to 
1. disable either of the buttons depending on the repo needs
2. choose which one is the default if both are enabled

**Comment by uaqben** [comment]
Same. Our team wish this were _already_ available. Hard to fathom why it isn't...

**Comment by spargo-nv** [comment]
I, too, would like the option to select `Rebase` as the default option for updating a branch.  
I had a team member request it to be the default option earlier today, which is how I found this thread, and several others asking for the same functionality.

**Comment by 64BitAsura** [comment]
@github disable update branch via `merge commit` when `allow merge commit` is disabled or something. I want update branch via rebase only if setting enabled ` allow merge via rebase only`  +1

**Comment by nocarroll-v8m** [comment]
+100, this feature would also greatly help our team as we are trying to move away from merge commits in our strategy.

**Comment by chirag-ema** [comment]
Please please add this feature. Seeing merge commits in our PRs disturbs a lot of future github actions (like creating releases and generating change logs)

**Comment by tompaz3** [comment]
It would be nice to be able disable "Update with rebase" completely as well.
We use commits signature verification and this feature is useless in our case, because when it does the rebase and force push, it creates new, unsigned commits which will fail the signature verification (it cannot sign the commits as it does not have the access to the private key, obviously).

Poeple still use the "Update with rebase" feature every once in a while and get confused why their PR checks fail and they have to do the interactive rebase locally to sign all the commits once again.

**Comment by jeffhole** [comment]
To add to the "disable" conversation, the "Update branch" button could have a section in Settings under General just like the Pull Request button:
 - [ ] allow merge commits
 - [ ] allow squash commits
 - [ ] allow rebasing commits

Update Branch button options:
 - [ ] allow update with merge commit
 - [ ] allow update with rebase

There should definitely be an option for selecting the default as well, for both buttons!

**Reply by erikacarvalho** [reply]
➕1️⃣

**Comment by mecampbellsoup** [comment]
What's the status of this, is GitHub even aware?

**Reply by sandstrom** [reply]
Don't know, but if you are a paying customer you can submit a feature request asking about it.

**Comment by repsejnworb** [comment]
Been awaiting this (and a "Update branch with rebase and merge into main"-button) for so many years.

Please Github

**Comment by WaleedAshraf** [comment]
Response from GitHub support as of today,

> Currently, Update with rebase as the default when updating a branch from the main branch option isn't available.
>  
> We have however heard this request from other customers as well. I have gone ahead and added your voice to our internal issue tracking this feature. I can not say if or when this feature might be added, but we'll make sure to document it in the [GitHub Changelog](https://github.blog/changelog/) or the [GitHub Blog](https://github.blog/) if it becomes available.

**Comment by 0cjs** [comment]
Here's another vote to add this.

**Comment by benjamin-brabant** [comment]
Here's another vote to add this.

**Reply by sandstrom** [reply]
If you are paying for Github, write their support and ask for it. Just writing it here will not make much of a difference.

**Comment by NathanBraslavski** [comment]
Count my vote as well, I will also write support.

**Comment by codella** [comment]
This would be very nice to have 🙏🏽 please, make it happen!

**Comment by repsejnworb** [comment]
I feel like over the years we've added thousands upon thousands of votes for this feature, but for the sake of it - here is my vote added (again).

**Comment by pieeri** [comment]
Adding our voice here as a very nice way and feature for enforcing such a development process at an organisation level!

**Comment by ankon** [comment]
Our workflow is to "rebase" branches in PRs, so that the history in the branch itself stays quite linear, and then create explicit merge commits (without squashing) when the PR branch gets merged into the mainline. 

It would be very helpful for us indeed if we could org-wide require the "Update branch" button to always use "rebase" as strategy.

**Reply by csi-lund** [reply]
org but also repo level configurable as well with the rules as a sub setting of requiring branches to be up to date

**Comment by rathpc** [comment]
It has been nearly 3 years since this issue was first raised, is there any movement on this at all?

**Reply by repsejnworb** [reply]
You should see the original original original xD

I've followed https://github.com/isaacs/github/issues/88#issuecomment-415875273 since it's inception, which was the first ask for something similar.

**Comment by ptrkstr** [comment]
Would love this feature as well 🥹

**Comment by gingters** [comment]
We would need this too. We want to explicitly FORBID updating the branch with a merge, ONLY rebase should be allowed.

**Comment by robertIsaac** [comment]
FYI adding a rule that require liner history, automatically disable the merge button and rebase become the only available option

**Reply by pmcgaley** [reply]
I tried this exact approach in the past and it looked like it was going to work, but hitting the button gave me this:

<img width="300" alt="there was a problem generating the rebase commit" src="https://github.com/user-attachments/assets/81233539-af86-4f60-ba7e-62ed2ee393f9" />

I just tried again and now it gives me this:

<img width="595" alt="Update branch attempt failed. Repository rule violations found. This branch must not contain merge commits" src="https://github.com/user-attachments/assets/038b7392-081d-4b23-8042-b9265f146592" />

I assume it's seeing that `main` has merge commits (the whole point of "semi-linear" history) and refusing to rebase the branch onto it. (I have `main` explicitly excluded from the rule)

**Reply by robertIsaac** [reply]
yeah well you need to clean the main branch first from merge commits
I used it in a new repo and it was fine but for old repos it's tricky

what you can do is to automated the rebase on first commit and to use `--ours` for example, this will remove all merge commits but the result won't match 100%
next use a tool to compare the main with remote/main (I used webstorm) and manually get all the changes in a new commit and force push to main
this will cause a big mess in the repo, if you decided to go this way ask all team members to delete and clone again the repo

**Reply by pmcgaley** [reply]
Sorry, crossed wires. This sounds like it would work if you're happy using purely linear history on main, but we (and some others on this thread) intentionally use semi-linear history - i.e. a merge commit for each PR, but PRs are always rebased before merging. So your suggestion wouldn't work for us, but may help others, thanks

**Reply by robertIsaac** [reply]
yes unfortunately it needs to be 100% liner for it to work
it's really great for new repos, but for old ones I have no idea what to do

**Reply by 0cjs** [reply]
I wonder if things might be tweaked so we could add a rule that since that the history must be linear since a given commit on a branch.

**Reply by sandstrom** [reply]
@robertIsaac It isn't only about old repos. Many new repos also want 'semi-linear merge' as the preferred strategy.

**Reply by robertIsaac** [reply]
> @robertIsaac It isn't only about old repos. Many new repos also want 'semi-linear merge' as the preferred strategy.

yeah I understand there can be some different use cases depending on different branching strategy

**Reply by csi-lund** [reply]
Thanks for raising the above points this is exactly what we need as well.

**Reply by dannysauer** [reply]
I have "require linear history" in a separate ruleset from the "required PR checks" (for automation purposes), and I get a merge option.  The behavior does not change when adding "require linear history" in the same ruleset as the required checks.  Does it only work for people using the old branch protection rules?

**Comment by alex-haas-bosch** [comment]
We would need this too.

**Comment by damusnet** [comment]
With the new Merge UI, the trick of adding a rule that requires a linear history, to automatically disable the merge button, and having rebase become the only available option, does not seem to work anymore. Worse, it appears that the Rebase option is not available at all anymore, as the dropdown is completely gone.

| <img width="826" alt="image" src="https://github.com/user-attachments/assets/38f83e24-66ab-4b93-8232-96f7f05d6fa1" /> | <img width="819" alt="image" src="https://github.com/user-attachments/assets/c1bf3158-998a-4e8e-a903-ba1b99600eda" /> | 
|--|--|

Edit: mmmh, maybe something else is happening, because if you remove the requirement for linear history, the dropdown comes back. Feels like a bug somewhere.

<img width="822" alt="image" src="https://github.com/user-attachments/assets/b2c5731a-783c-4d5c-b526-a237de8e88fa" />

**Reply by uaqben** [reply]
![](https://media.tenor.com/xUdqC5nsiNUAAAAM/falling-dead.gif)

**Reply by kragoth** [reply]
And now the work around doesn't seem to work for me anymore. *sigh*
My team now have to rebase our PRs locally and force push. There's no way for me to do it via the PR screen at all now.


Edit: nevermind, I got the workaround working again.

**Reply by aaron-riact** [reply]
My current work around is to let devs do whatever they want on branches, but require that PRs themselves are up to date and enforce that they have no merges in a github action.  Perhaps there is a better way of achieving this?

```
      - name: Check out this repo
        uses: actions/checkout@v4
        with:
          path: repo
          fetch-depth: 0

      - name: Ensure PR has linear history
        if: github.event_name == 'pull_request'
        working-directory: ./repo
        run: >
          if [[ -n $(git rev-list origin/$GITHUB_BASE_REF..origin/$GITHUB_HEAD_REF --merges) ]]; then
            echo "::error::PR contains merges"
            exit 1
          fi
```

**Comment by fgmehlin** [comment]
For bigger teams it can get pretty messy if we start updating with both rebasing and merging. Teams usually stick to one option only, so that seems to be pretty much a must have.

Gitlab offers this for a long time already. I sincerely hope this is added very soon. 

Here's another strong vote for this.

**Comment by davsv** [comment]
I'd like a feature that DISABLES the "Update with rebase" and ONLY uses "Update with merge commit". I have disabled deletion and force pushing for all branches but the "Update with rebase" still shows up.

**Reply by 0cjs** [reply]
This has piqued my curiosity; what goes wrong when someone updates their development branch with a rebase rather than a merge?

**Reply by dannysauer** [reply]
Because rebase works by replaying commits, you lose the original signatures doing a rebase on GitHub. I suppose that's a negative.

**Reply by ptrkstr** [reply]
> This has piqued my curiosity; what goes wrong when someone updates their development branch with a rebase rather than a merge?

@0cjs GitHub recreates the commits meaning they have different commit hashes

**Reply by 0cjs** [reply]
So those reasons make sense to me for merges of dev branches into main but here we're talking about the other direction: updating a dev branch with new commits from main. What's the advantage of preserving commits on a dev branch? (The disadvantage seems obvious; if after review you fix a mistake, you end up preserving the mistake and the fix on main, rather than, when you merge to main, having just the correct version there. The same goes for, e.g., a rename from a new refactoring on main: if someone renames `a` to `b`, your newly generated commits from a rebase are all using `b` from the start, rather than using `a` and then having a merge commit that changes the `a`s to `b`s.)

**Reply by dannysauer** [reply]
@0cjs the op just said branches, not developer branches. ;) There are development strategies which aren't trunk based, where more branches than Main are either long lived or otherwise aren't individual developer working spaces. :)

This is also primarily talking about the "update Branch with rebase" button on GitHub, not telling people to avoid rebasing locally. Though preventing force pushes would also block local rebases. 🤷‍♂️ I can definitely imagine a world where blanket policy would be to never rewrite history. For example, I have a gitops IaC repository where the content is mostly automatically generated; "developers" are usually programs and I actually need to record *all* history for change management / compliance / audit purposes.

For what it's worth, I don't disagree. Just listing some other scenarios. :)

**Reply by 0cjs** [reply]
@dannysauer No, the OP was talking specifically about the "Update with rebase" and "Update with merge commit" buttons, which update a dev branch from the main branch. And note that in his case developers cannot rebase locally either, or to be more precise, they could do that but they can't push up those changes because he's disabled force pushing for all branches.

And I am familiar with using Git for compliance, etc., with automatically generated content; this just doesn't sound like that kind of situation at all.

FWIW, I do agree that being able to use the "Update with rebase" button on branches that are not allowed to be force-pushed is probably wrong behaviour, regardless of the reasons someone disabled force-push.

**Comment by allisonphillips** [comment]
# I've sacrificed my Friday night to figure out all the things...

## Repository Settings
- Check `Allow rebase merging` and `Always suggest updating pull request branches`
  <img width="731" alt="Screenshot 2025-06-13 at 6 11 35 PM" src="https://github.com/user-attachments/assets/16c8b1ef-524e-4a2c-8136-97beaf2b6d23" />


## Rulesets
### Go to `Rules` &rarr; `Rulesets`

### Create a `branch` ruleset
<img width="195" alt="Screenshot 2025-06-13 at 6 16 07 PM" src="https://github.com/user-attachments/assets/4aae8560-9d2d-4934-b230-0127ffb22afb" />

### Target ALL branches
<img width="722" alt="Screenshot 2025-06-13 at 6 15 42 PM" src="https://github.com/user-attachments/assets/3db58bf7-8546-4d91-96a1-b137f3606bbe" />

### Check the box for `Require linear history`
<img width="410" alt="Screenshot 2025-06-13 at 6 15 50 PM" src="https://github.com/user-attachments/assets/5bc98aff-3e92-47f3-b84b-2baaafbeb805" />

### Save & Activate it

### Exported Rule
If you export your saved rule, it should look something like this:
```json
{
  "name": "Linear History - Rebase",
  "target": "branch",
  "source_type": "Repository",
  "enforcement": "active",
  "conditions": {
    "ref_name": {
      "include": [
        "~ALL"
      ]
    }
  },
  "rules": [
    {
      "type": "required_linear_history"
    }
  ],
  "bypass_actors": []
}
```

## PR Experience
When a branch is out of date, it will show a single rebase button to resync the branch:
<img width="787" alt="Screenshot 2025-06-13 at 10 17 57 PM" src="https://github.com/user-attachments/assets/31ebd17e-82f3-44c6-a9e8-374a36ca9128" />

If it cannot be rebased, it will list the files that need merge conflict resolutions.

--- 

# 🤬 I enabled the linear branch history rule and **now I can't push new expletive branches** 👇🏾 

If your default branch's history is non-linear, the branch rule will prevent you from pushing a new branch (even if your new commits are linear and you are rebased on master).

## Today's the day to rename `master` to `main`!
- Create your new branch off of the old default branch, [rebase --root], and push it up:
  ```
  git checkout -b main origin/master
  git rebase --root
  git push origin head
  ```
  Sure, you could just `rebase --root` master and force push it up, but let's assume you are like me in being both risk-averse and want to err on the side of being meticulous for the purposes of compliance audits and general continued employment.
- Confirm that you correctly duplicated master:
  `http://github.com/$ORGANIZATION/$REPO/compare/main..master`
  <img width="1131" alt="Screenshot 2025-06-13 at 9 00 21 PM" src="https://github.com/user-attachments/assets/75d817d4-f926-4327-b6bc-5a4f77ca8390" />

- Switch your default branch in the repository settings (there's an &hArr;-ish icon to the right of the branch name, that will let you switch to a different default branch)
  - Update your branch protection rule to _exclude_ `master-non-linear-archive` (or it will prevent you from renaming master)
  - Go into your repository's branches and rename `master` to `master-non-linear-archive`
  - Create a legacy branch protection rule to lock the branch
- Change the tracking branch on existing PRs to main (edit the PR description and it will let you switch the branch), rebase each branch on origin/main and force push back to origin (or make your team do it if you are some sort of cruel overlord)
- Update any workflows or other docu that might reference your old default branch name.
- Set head locally:
  ```
  git branch -u origin/main main
  git remote set-head origin -a
  ```


😌 

[rebase --root]: https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---root

**Reply by 0cjs** [reply]
Can you add an image of the new drop-down you now get?

**Reply by allisonphillips** [reply]
> Can you add an image of the new drop-down you now get?

It's just a button, but I added it.

I also ran into another fun new problem, so there is a new section for `🤬 I enabled the linear branch history rule and now I can't push new expletive branches 👇🏾`

**Reply by gcv-epalmer** [reply]
Thank you for your service 🙏

**Reply by thornycrackers** [reply]
Bless you for writing all this up

**Reply by sidane** [reply]
Thank you so much for this ❤️

**Comment by vtjballeng** [comment]
> mal to always have to pick the rebase option from a drop down menu.
> 
> It would be ideal if there were a repository level setting to configure which option is the default for that button.

Same here! We want a way to lock out the default update branch merge commit behavior in favor of rebase.

**Comment by roger-watkins** [comment]
I would love a way to just disable this button as both merge and rebase in the UI wrecks our PRs.

- Merge update prevents PR merging due to linear-history (which I really, really want to keep)
- Rebase update wrecks commit signing.

Given that both options are useless for us why can I not just turn it off (Always suggest updating pull request branches - does nothing).

**Reply by 0cjs** [reply]
Yes, I've always been annoyed by the way when you could just fast-forward GH insists on rewriting the commits anyway.

**Reply by Primajin** [reply]
Isn't there a setting under General:

Control how and when users are prompted to update their branches if there are new changes available in the base branch.

☑️ Always suggest updating pull request branches
Whenever there are new changes available in the base branch, present an “update branch” option in the pull request.

This should allow you to turn off the feature, right?

**Reply by 0cjs** [reply]
No, that's orthogonal. Once the pull request branch is updated, it's directly on head and could be fast-forwarded (e.g., with `git push origin @:main` when you're on the branch), but if you use the button in the GitHub API it will still rewrite all the commits, putting new but equivalent versions of the branch's commits on to `main`.

**Reply by Primajin** [reply]
Maybe then I have a misunderstanding. @roger-watkins asked for a way to disable the button and I thought disabling the option under general would solve that.

**Reply by 0cjs** [reply]
Ah, I see now; sorry for misinterpreting you. I was discussing in this thread within the issue what @roger-watkins raised, which is how the rewrite is done. but for the original issue the problem would not solved by disabling the button, but by disabling the option "under" the button to merge, leaving the option "under" the button to rebase.

**Comment by Sirivasv** [comment]
> It would be ideal if there were a repository level setting to configure which option is the default for that button.

Keeping this topic alive sharing that our team would also benefit from this! Better yet to allow only one of them.

**Comment by MichaelKetting** [comment]
I'm also still holding out hope that this feature will be implemented. It should a) hopefully be an inexpensive change and is b) a great way to either differentiate or catch up to the competition.
